From 9220c2a77eda3df37a84486ad722f1ad0985d8e7 Mon Sep 17 00:00:00 2001
From: Kenny2github <kenny2minecraft@gmail.com>
Date: Tue, 13 Apr 2021 11:48:24 +0100
Subject: [PATCH] SECURITY: Mitigate MITM (GHSA-gvpg-23fh-8g75)

Prepend an anti-phishing message to the code itself warning about source
Add instructions to delete the comment afterwards to protect privacy
---
 i18n/en.json                              | 2 ++
 i18n/qqq.json                             | 2 ++
 includes/common/login.php                 | 1 +
 includes/special/SpecialScratchOAuth2.php | 3 ++-
 4 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/i18n/en.json b/i18n/en.json
index 509d32b..665e9e7 100644
--- a/i18n/en.json
+++ b/i18n/en.json
@@ -7,6 +7,8 @@
 	"soa2-scratch-username": "Scratch username:",
 	"soa2-invalid-username": "Invalid username: $1",
 	"soa2-vercode-explanation": "Please go to $1 and comment the code below:",
+	"soa2-vercode-explanation2": "For added privacy, delete your comment immediately after logging in here.",
+	"soa2-vercode": "Copy this paragraph (including both the code and this message). Only post this code if it came from {{SERVER}} | $1",
 	"soa2-your-profile": "your profile",
 	"soa2-next": "Next",
 	"soa2-login": "I have done so",
diff --git a/i18n/qqq.json b/i18n/qqq.json
index 477ad72..7e3e322 100644
--- a/i18n/qqq.json
+++ b/i18n/qqq.json
@@ -7,6 +7,8 @@
 	"soa2-scratch-username": "Label for Scratch username field",
 	"soa2-invalid-username": "Error message for invalid username",
 	"soa2-vercode-explanation": "Instructions for user verification. $1 is the \"your profile\" link.",
+	"soa2-vercode-explanation2": "Instructions to delete the comment afterwards.",
+	"soa2-vercode": "A message prepended to the actual code to prevent man-in-the-middle attacks. $1 is the code.",
 	"soa2-your-profile": "Link text for \"your profile\"",
 	"soa2-next": "Next button text",
 	"soa2-login": "Login button text",
diff --git a/includes/common/login.php b/includes/common/login.php
index 06fa792..e5f5bbc 100644
--- a/includes/common/login.php
+++ b/includes/common/login.php
@@ -14,6 +14,7 @@ public static function gen_code( $session ) {
 		if (!$session->exists( 'soa2_scratch_code' )) {
 			$code = chunk_split(hash('sha256', random_bytes(32)), 5, ':');
 			$code = substr($code, 0, strlen($code) - 1); // chop off last colon
+			$code = wfMessage('soa2-vercode', $code)->text();
 			$session->set( 'soa2_scratch_code', $code );
 			$session->save();
 		}
diff --git a/includes/special/SpecialScratchOAuth2.php b/includes/special/SpecialScratchOAuth2.php
index 2aafe15..5461d0e 100644
--- a/includes/special/SpecialScratchOAuth2.php
+++ b/includes/special/SpecialScratchOAuth2.php
@@ -120,8 +120,9 @@ public function loginForm( WebRequest $request ) { // Step 13
 			wfMessage('soa2-vercode-explanation')->rawParams($profile)->parse()
 		));
 		$out->addHTML(Html::rawElement('p', [], Html::element(
-			'code', [], $codes['code']
+			'textarea', [], $codes['code']
 		)));
+		$out->addWikiMsg('soa2-vercode-explanation2');
 		$out->addHTML(Html::rawElement('p', [], Html::submitButton(
 			wfMessage('soa2-login')->plain(), []
 		)));
